test/persistent-testdrive-example/mzcompose.py

```diff
@@ -0,0 +1,30 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from pathlib import Path
+from typing import List
+
+from materialize.mzcompose import Composition, WorkflowArgumentParser
+from materialize.mzcompose.services import Materialized, Testdrive
+
+td = Testdrive(persistent=True)
+
+SERVICES = [Materialized(), td]
+
+
+def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+    c.start_and_wait_for_tcp(services=["materialized"])
+    c.wait_for_materialized("materialized")
+
+    c.up("testdrive-svc")
+
+    for i in range(1, 8):
+        td.exec(c, f"test{i}.td")
```

Comment:
A bit unfortunate that we have to break with the API style. Like, ideally this would be `c.exec("testdrive-svc", f"test{i}.td")`.

Reply:
Oh, we can! I'd be inclined to make `persistent=True` an option you pass to `Composition.up`, rather than something that we need to invent for each service. Here's an amalgamated diff that includes both the diff from the next comment _and_ this idea:

```diff
diff --git a/misc/python/materialize/mzcompose/__init__.py b/misc/python/materialize/mzcompose/__init__.py
index a84e297be..3f5078ce0 100644
--- a/misc/python/materialize/mzcompose/__init__.py
+++ b/misc/python/materialize/mzcompose/__init__.py
@@ -372,39 +372,6 @@ class Composition:
                 print(e.stdout)
             raise UIError(f"running docker-compose failed (exit status {e.returncode})")
 
-    def invoke_docker(
-        self, *args: str, capture: bool = False
-    ) -> subprocess.CompletedProcess:
-        """Invoke `docker` directly
-
-        Args:
-            args: The arguments to pass to `docker`.
-            capture: Whether to capture the child's stdout stream.
-        """
-
-        if not self.silent:
-            print(f"$ docker {' '.join(args)}", file=sys.stderr)
-
-        stdout = None
-        if capture:
-            stdout = subprocess.PIPE
-
-        try:
-            return subprocess.run(
-                [
-                    "docker",
-                    *args,
-                ],
-                close_fds=False,
-                check=True,
-                stdout=stdout,
-                text=True,
-            )
-        except subprocess.CalledProcessError as e:
-            if e.stdout:
-                print(e.stdout)
-            raise UIError(f"running docker failed (exit status {e.returncode})")
-
     def port(self, service: str, private_port: Union[int, str]) -> int:
         """Get the public port for a service's private port.
 
@@ -609,22 +576,23 @@ class Composition:
         )
 
     def exec(
-        self, service: "Service", *args: str, detach: bool = False
+        self, service: str, *args: str, detach: bool = False
     ) -> subprocess.CompletedProcess:
         """Execute a one-off command in a service's running container
-        Delegates to `docker exec` as `docker-compose exec` appears to do nothing
+
+        Delegates to `docker-compose exec`.
+
         Args:
             service: The service whose container will be used
             command: The command to run
             args: Arguments to pass to the command
         """
 
-        container = "_".join([self.name, service.name, "1"])
-
-        return self.invoke_docker(
+        return self.invoke(
             "exec",
             *(["--detach"] if detach else []),
-            container,
+            service,
+            *self.compose["services"][service]["entrypoint"],
             *args,
         )
 
@@ -642,7 +610,7 @@ class Composition:
             ):
                 self.invoke("pull", service)
 
-    def up(self, *services: str, detach: bool = True) -> None:
+    def up(self, *services: str, detach: bool = True, persistent: bool = False) -> None:
         """Build, (re)create, and start the named services.
 
         Delegates to `docker-compose up`. See that command's help for details.
@@ -650,9 +618,23 @@ class Composition:
         Args:
             services: The names of services in the composition.
             detach: Run containers in the background.
+            persistent: Replace the container's entrypoint and command with
+                `sleep infinity` so that additional commands can be scheduled
+                on the container with `Composition.exec`.
         """
+        if persistent:
+            old_compose = copy.deepcopy(self.compose)
+            for service in self.compose["services"].values():
+                service["entrypoint"] = ["sleep", "infinity"]
+                service["command"] = []
+            self._write_compose()
+
         self.invoke("up", *(["--detach"] if detach else []), *services)
 
+        if persistent:
+            self.compose = old_compose
+            self._write_compose()
+
     def down(self, destroy_volumes: bool = True) -> None:
         """Stop and remove resources.
 
diff --git a/misc/python/materialize/mzcompose/services.py b/misc/python/materialize/mzcompose/services.py
index be58fa788..b36035d37 100644
--- a/misc/python/materialize/mzcompose/services.py
+++ b/misc/python/materialize/mzcompose/services.py
@@ -495,7 +495,6 @@ class Testdrive(Service):
         volume_workdir: str = ".:/workdir",
         propagate_uid_gid: bool = True,
         forward_buildkite_shard: bool = False,
-        persistent: bool = False,
     ) -> None:
         if environment is None:
             environment = [
@@ -562,7 +561,7 @@ class Testdrive(Service):
             name=name,
             config={
                 "mzbuild": mzbuild,
-                "entrypoint": ["sleep", "infinity"] if persistent else entrypoint,
+                "entrypoint": entrypoint,
                 "environment": environment,
                 "volumes": volumes,
                 "propagate_uid_gid": propagate_uid_gid,
@@ -570,9 +569,6 @@ class Testdrive(Service):
             },
         )
 
-    def exec(self, c: "Composition", *args: str) -> None:
-        c.exec(self, *(self.entrypoint), *args)
-
 
 class TestCerts(Service):
     def __init__(
diff --git a/test/persistent-testdrive-example/mzcompose.py b/test/persistent-testdrive-example/mzcompose.py
index fec2442b4..afbfcc91a 100644
--- a/test/persistent-testdrive-example/mzcompose.py
+++ b/test/persistent-testdrive-example/mzcompose.py
@@ -13,18 +13,16 @@ from typing import List
 from materialize.mzcompose import Composition, WorkflowArgumentParser
 from materialize.mzcompose.services import Materialized, Testdrive
 
-td = Testdrive(persistent=True)
-
-SERVICES = [Materialized(), td]
+SERVICES = [Materialized(), Testdrive()]
 
 
 def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
     c.start_and_wait_for_tcp(services=["materialized"])
     c.wait_for_materialized("materialized")
 
-    c.up("testdrive-svc")
+    c.up("testdrive-svc", persistent=True)
 
     for i in range(1, 8):
-        td.exec(c, f"test{i}.td")
+        c.exec("testdrive-svc", f"test{i}.td")
 
     c.rm("testdrive-svc")
```